[BPF] introduce `__attribute__((bpf_fastcall))`

eddyz87 opened the pull request:
This commit introduces attribute bpf_fastcall to declare BPF functions that do not clobber some of the caller saved registers (R0-R5).

The idea is to generate the code complying with generic BPF ABI, but allow compatible Linux Kernel to remove unnecessary spills and fills of non-scratched registers (given some compiler assistance).

For such functions do register allocation as-if caller saved registers are not clobbered, but later wrap the calls with spill and fill patterns that are simple to recognize in kernel.

For example for the following C code:

     #define __bpf_fastcall __attribute__((bpf_fastcall))

     void bar(void) __bpf_fastcall;
     void buz(long i, long j, long k);

     void foo(long i, long j, long k) {
       bar();
       buz(i, j, k);
     }

First allocate registers as if:

    foo:
      call bar    # note: no spills for i,j,k (r1,r2,r3)
      call buz
      exit

And later insert spills fills on the peephole phase:

    foo:
      *(u64 *)(r10 - 8) = r1;  # Such call pattern is
      *(u64 *)(r10 - 16) = r2; # correct when used with
      *(u64 *)(r10 - 24) = r3; # old kernels.
      call bar
      r3 = *(u64 *)(r10 - 24); # But also allows new
      r2 = *(u64 *)(r10 - 16); # kernels to recognize the
      r1 = *(u64 *)(r10 - 8);  # pattern and remove spills/fills.
      call buz
      exit

The offsets for generated spills/fills are picked as minimal stack offsets for the function. Allocated stack slots are not used for any other purposes, in order to simplify in-kernel analysis.

Corresponding functionality had been merged in Linux Kernel as [this](https://lore.kernel.org/bpf/172179364482.1919.9590705031832457529.git-patchwork-notify@kernel.org/) patch set (the patch assumed that `no_caller_saved_regsiters` attribute would be used by LLVM, naming does not matter for the Kernel).

eddyz87 committed:
[BPF] introduce __attribute__((bpf_fastcall))

This commit introduces attribute bpf_fastcall to declare BPF functions
that do not clobber some of the caller saved registers (R0-R5).

The idea is to generate the code complying with generic BPF ABI,
but allow compatible Linux Kernel to remove unnecessary spills and
fills of non-scratched registers (given some compiler assistance).

For such functions do register allocation as-if caller saved registers
are not clobbered, but later wrap the calls with spill and fill
patterns that are simple to recognize in kernel.

For example for the following C code:

   #define __bpf_fastcall __attribute__((bpf_fastcall))

   void bar(void) __bpf_fastcall;
   void buz(long i, long j, long k);

   void foo(long i, long j, long k) {
     bar();
     buz(i, j, k);
   }

First allocate registers as if:

   foo:
     call bar    # note: no spills for i,j,k (r1,r2,r3)
     call buz
     exit

And later insert spills fills on the peephole phase:

   foo:
     *(u64 *)(r10 - 8) = r1;  # Such call pattern is
     *(u64 *)(r10 - 16) = r2; # correct when used with
     *(u64 *)(r10 - 24) = r3; # old kernels.
     call bar
     r3 = *(u64 *)(r10 - 24); # But also allows new
     r2 = *(u64 *)(r10 - 16); # kernels to recognize the
     r1 = *(u64 *)(r10 - 8);  # pattern and remove spills/fills.
     call buz
     exit

The offsets for generated spills/fills are picked as minimal stack
offsets for the function. Allocated stack slots are not used for any
other purposes, in order to simplify in-kernel analysis.

efriedma-quic commented:
Is there some reason this is an attribute, and not a calling convention, at the IR level?

eddyz87 commented:
> Is there some reason this is an attribute, and not a calling convention, at the IR level?

Thought about it and decided against that, but I agree that this is an option, my reasoning below.
From the semantic point of view the difference between current attribute implementation and calling convention would be whether the code below reports be an error:

```c
    void (*ptr1)(void) __bpf_fastcall;
    void (*ptr2)(void);
    void foo(void) {
      ptr2 = ptr1; // is this an error?
    }
```

From the Kernel point of view it really is not, as this "fast call" is an optimization hint, that could be safely ignored by BPF jit. Plus desire to keep clang front-end changes to the minimum. Also, the argument made for `preserve_caller_saved_registers` [here](https://reviews.llvm.org/D22045#494112) sort-of makes sense for `bpf_fastcall` as well.

> And why would it not be named __attribute__((fastcall)) when targeting BPF

I wanted to avoid potential confusion, as `bpf_fastcall` operates in a manner completely different from what `fastcall` for x86 does (first two parameters in `ecx`, `edx`, other parameters on stack).

yonghong-song reviewed (approved):
I also tried a few examples and it looks good to me. Only a few minor comments.

eddyz87 commented on clang/test/CodeGen/bpf-attr-bpf-fastcall-1.c:
```diff
@@ -0,0 +1,24 @@
+// CHECK:   ret void
+
+// CHECK: declare void @test() #[[ptr_attr:[0-9]+]]
+// CHECK: attributes #1 = { {{.*}}"bpf_fastcall"{{.*}} }
```
yes, thank you for catching this

eddyz87 commented on clang/test/CodeGen/bpf-attr-bpf-fastcall-1.c:
```diff
@@ -0,0 +1,24 @@
+
+// CHECK: declare void @test() #[[ptr_attr:[0-9]+]]
+// CHECK: attributes #1 = { {{.*}}"bpf_fastcall"{{.*}} }
+// CHECK: attributes #[[test_attr]] = { {{.*}}"bpf_fastcall"{{.*}} }
```
only for `#1`, I'll remove `{{.*}}` for `test_attr`.

eddyz87 commented on clang/test/CodeGen/bpf-attr-bpf-fastcall-1.c:
```diff
@@ -0,0 +1,24 @@
+}
+
+// CHECK: @ptr = global ptr null
+// CHECK: define {{.*}} @foo()
```
Will change

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -596,6 +600,90 @@ bool BPFMIPreEmitPeephole::adjustBranch() {
+    unsigned LiveCallerSavedRegs;
+    if (!MI.isCall())
+      goto NextInsn;
+    LiveCallerSavedRegs = 0;
```
Ok

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -596,6 +600,90 @@ bool BPFMIPreEmitPeephole::adjustBranch() {
+  bool Changed = false;
+  for (MachineBasicBlock &BB : *MF) {
+    collectBPFFastCalls(TRI, LiveRegs, BB, Calls);
+    Changed |= !Calls.empty();
```
In `collectBPFFastCalls` the calls w/o spills/fills are not added to the `Calls`.

yonghong-song replied:
Okay, I see. You are right. Thanks for clarification!

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -596,6 +600,90 @@ bool BPFMIPreEmitPeephole::adjustBranch() {
+  for (MachineInstr &MI : llvm::reverse(BB)) {
+    unsigned LiveCallerSavedRegs;
+    if (!MI.isCall())
+      goto NextInsn;
```
Ok

eddyz87 committed:
styling fix: remove goto in collectBPFFastCalls

eddyz87 committed:
test fix: use [[names]] for attributes in bpf-attr-bpf-fastcall-1

Also change attribute names to be less confusing.

eddyz87 commented:
@AaronBallman , @efriedma-quic , could you please check my last comment [here](https://github.com/llvm/llvm-project/pull/101228#issuecomment-2263848414)?

AaronBallman commented:
> > Is there some reason this is an attribute, and not a calling convention, at the IR level?
> 
> Thought about it and decided against that, but I agree that this is an option, my reasoning below. From the semantic point of view the difference between current attribute implementation and calling convention would be whether the code below reports be an error:
> 
> ```c
>     void (*ptr1)(void) __bpf_fastcall;
>     void (*ptr2)(void);
>     void foo(void) {
>       ptr2 = ptr1; // is this an error?
>     }
> ```
> 
> From the Kernel point of view it really is not, as this "fast call" is an optimization hint, that could be safely ignored by BPF jit. Plus desire to keep clang front-end changes to the minimum. Also, the argument made for `preserve_caller_saved_registers` [here](https://reviews.llvm.org/D22045#494112) sort-of makes sense for `bpf_fastcall` as well.

Doesn't that kind of defeat the purpose of the calling convention? (A function designator decays into a function pointer basically any time you mention it.)

> This commit introduces attribute bpf_fastcall to declare BPF functions that do not clobber some of the caller saved registers (R0-R5).

If the attribute can be dropped, then the caller side has to assume the caller-saved registers are still going to be clobbered even in the presence of the attribute, so it always has to save and restore those registers, doesn't it?

eddyz87 commented:
@AaronBallman,

> > ```c
> >     void (*ptr1)(void) __bpf_fastcall;
> >     void (*ptr2)(void);
> >     void foo(void) {
> >       ptr2 = ptr1; // is this an error?
> >     }
> > ```
> ...
> Doesn't that kind of defeat the purpose of the calling convention? (A function designator decays into a function pointer basically any time you mention it.)
>
> > This commit introduces attribute bpf_fastcall to declare BPF functions that do not clobber some of the caller saved registers (R0-R5).
>
> If the attribute can be dropped, then the caller side has to assume the caller-saved registers are still going to be clobbered even in the presence of the attribute, so it always has to save and restore those registers, doesn't it?

If attribute is dropped, then yes. E.g. if function is called through `ptr2` from the example above, the register allocator would assume that all caller saved registers are clobbered. However, the generated code is still correct, nothing will break or behave unexpectedly. Worst case, some performance would be left on a table. That's why I'm sort-of inclined to say that this is not an error.

eddyz87 commented:
Moreover, BPF only has one calling convention at the moment: all parameters are passed in registers (R1-R5), return value in R0. But assuming a new calling convention would be added one day, e.g. some parameters passed on stack, the feature in question would still be applicable => this behaves more like an attribute, not a calling convention (same as existing `no_preserve_caller_saved_registers`).

AaronBallman commented:
> If attribute is dropped, then yes. E.g. if function is called through `ptr2` from the example above, the register allocator would assume that all caller saved registers are clobbered. However, the generated code is still correct, nothing will break or behave unexpectedly. Worst case, some performance would be left on a table. That's why I'm inclined to say that this is not an error.

But a mismatch can still potentially result in a miscompilation, right? e.g., you have a function in a header file with the `bfp_fastcall` attribute on it. The definition of the function is compiled into a library with Clang 16 and ignores the unknown attribute, so the callee will clobber registers. But the declaration is used by an application compiled with Clang 20 and assumes the attribute means the callee won't clobber registers, so doesn't generate the save/restore code. When the call resolves, the registers are clobbered unexpectedly, right? But if the attribute was part of the function type, this program presumably would not link due to type mismatch.

eddyz87 commented:
@AaronBallman,

> But a mismatch can still potentially result in a miscompilation, right? e.g., you have a function in a header file with the `bfp_fastcall` attribute on it. The definition of the function is compiled into a library with Clang 16 and ignores the unknown attribute, so the callee will clobber registers. But the declaration is used by an application compiled with Clang 20 and assumes the attribute means the callee won't clobber registers, so doesn't generate the save/restore code. When the call resolves, the registers are clobbered unexpectedly, right? But if the attribute was part of the function type, this program presumably would not link due to type mismatch.

Not quite, the feature is a bit unusual and pursues several goals:
- allow bpf programs run on new kernels (those that support `bpf_fastcall`) to gain some performance by clobbering less registers;
- allow same code to run unmodified on old kernels (those that do not support `bpf_fastcall`);
- allow `bpf_fastcall` only for "builtin" functions, implemented by kernel, not written in BPF;
- allow new kernels to provide different sets of `bpf_fastcall` builtin functions depending on configuration and architecture;
- do not complicate kernel side implementation much (read as: do not do register allocation in kernel jit).

Suppose there is a function `void foo(void)` marked with `bpf_fastcall` and the following IR:

    %v = ... define %v somehow ...
    call void @foo
    ... use %v somehow ...

In such a case:
- only `@foo` prototype is visible to BPF program, not it's body;
- for the BPF backend the attribute serves as a hint for register allocation, the BPF code like below might be generated:

```
    r1 = ... definition of %v ...
    fp[-32] = r1
    call foo
    r1 = fp[-32]
    ... use r1 as %v ...
```

Kernels supporting `bpf_fastcall` for function `foo` would remove `r1` spill and fill, kernels not supporting `bpf_fastcall` would just work as before.

eddyz87 commented:
@AaronBallman, @efriedma-quic, could you please comment?
We are eager to use this feature on the BPF side (e.g. [here](https://lore.kernel.org/bpf/CAADnVQJDki9GCxDAaGJWb+HrKT2EnzYXM8K3238XxPtHkhU0Ag@mail.gmail.com/) and landing it to the `main` would simplify the cooperation.

AaronBallman commented:
Thank you (all) for the explanations, that was helpful -- it addressed my concerns with the design.

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2189,6 +2189,14 @@ def BTFTypeTag : TypeAttr {
+def BPFFastCall : InheritableAttr,
+                  TargetSpecificAttr<TargetBPF> {
+  let Spellings = [GCC<"bpf_fastcall">];
+  let Subjects = SubjectList<[FunctionLike]>;
```
Do you have to care about Objective-C methods or C++ member functions?

eddyz87 replied:
No, only focus on C at the moment.

AaronBallman replied:
Excellent!

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2189,6 +2189,14 @@ def BTFTypeTag : TypeAttr {
 
+def BPFFastCall : InheritableAttr,
+                  TargetSpecificAttr<TargetBPF> {
+  let Spellings = [GCC<"bpf_fastcall">];
```
Does GCC support this attribute with this spelling? If not, this should use the `Clang` spelling instead of the `GCC` spelling.

eddyz87 replied:
GCC will have to catch-up, but the would also have to do so for `btf_{type,decl}_tag`. Changed to `Clang`.

AaronBallman commented on clang/test/Sema/bpf-attr-bpf-fastcall.c:
```diff
@@ -0,0 +1,7 @@
+__attribute__((bpf_fastcall)) int var; // expected-warning {{'bpf_fastcall' attribute only applies to functions and function pointers}}
+
+__attribute__((bpf_fastcall)) void func();
+__attribute__((bpf_fastcall(1))) void func_invalid(); // expected-error {{'bpf_fastcall' attribute takes no arguments}}
```
Can you also add a test applying it to a function pointer, with some tests showing that this attribute is not part of the type (so there's no pointer mismatches)?

If you have to care about C++ member functions, a test for that would be useful as well (and a codegen test too).

eddyz87 replied:
Sure, will do, thank you for the review

eddyz87 committed:
use `Clang` syntax instead of `GCC` for bpf_fastcall

eddyz87 committed:
adjust Sema test case to show that bpf_fastcall is not a type attr

As discussed in the pull request thread, it is not an error to assign
pointer with bpf_fastcall attribute to a pointer w/o such attribute
and vice-versa.
(Hence, no need to track bpf_fastcall as a part of the type).

yonghong-song commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2189,6 +2189,14 @@ def BTFTypeTag : TypeAttr {
+  let Subjects = SubjectList<[FunctionLike]>;
+  let Documentation = [BPFFastCallDocs];
+  let SimpleHandler = 1;
+}
```
If we want to enforce attribute only for C lang, we can add `let LangOpts = [COnly];` in the above.

eddyz87 replied:
Makes sense

eddyz87 replied:
done

eddyz87 committed:
mark bpf_fastcall as COnly attribute

AaronBallman reviewed (approved):
Clang bits LGTM; accepting but please wait to land until someone accepts the LLVM changes as well.

eddyz87 commented:
> Clang bits LGTM; accepting but please wait to land until someone accepts the LLVM changes as well.

@AaronBallman , thank you for the review!
On the LLVM side the changes are only for BPF backend and these were already approved by @yonghong-song, or do you have someone else in mind?

yonghong-song commented:
@AaronBallman the llvm side of change looks good to me!

AaronBallman commented:
> @AaronBallman the llvm side of change looks good to me!

Then I think we're good to land! @eddyz87 do you need someone to land on your behalf?

eddyz87 commented:
> Then I think we're good to land! @eddyz87 do you need someone to land on your behalf?

@AaronBallman, I have commit rights, will merge this change shortly, thank you!

eddyz87 commented:
There was a failure for expansive checks:
https://lab.llvm.org/buildbot/#/builders/187/builds/509

I reverted the commit, will investigate.

eddyz87 commented:
@AaronBallman, could you please help me with the procedural question:
I reverted the commit and current pull request is now closed, what should I do when I'll figure out the fix?
It looks like there is no option to re-open pull request, hence it looks like I will have to open a new pull request.

vvereschaka commented:
Hi @eddyz87 ,

there are few failed `fastcall` tests on `[llvm-clang-x86_64-expensive-checks-ubuntu](https://lab.llvm.org/buildbot/#/builders/187) also

https://lab.llvm.org/buildbot/#/builders/187/builds/509

* LLVM::bpf-fastcall-regmask-1.ll
* LLVM::bpf-fastcall-1.ll
* LLVM::bpf-fastcall-3.ll
* LLVM::bpf-fastcall-2.ll

eddyz87 commented:
Hi @vvereschaka ,

Yes, I receive email notifications, the commit was reverted an hour ago.
While at it, could you please help me understand current process, should I open a new pull request once I figure out the fix?

vvereschaka commented:
@eddyz87 ,

just create a new PR with the changes and make the cross references between this PR and the new one to let people easy find the original PR when necessary.

AaronBallman commented:
> @eddyz87 ,
> 
> just create a new PR with the changes and make the cross references between this PR and the new one to let people easy find the original PR when necessary.

Yup, this is the typical process (GitHub doesn't let you reopen a merged PR).

eddyz87 commented:
> > @eddyz87 ,
> > just create a new PR with the changes and make the cross references between this PR and the new one to let people easy find the original PR when necessary.
> 
> Yup, this is the typical process (GitHub doesn't let you reopen a merged PR).

Understood, thank you.
Will post the new pr.
